### ISSUE TITLE
[Feature request] Custom SSL local path

### ISSUE THREAD

**Issue by eithe** [jc21/nginx-proxy-manager#87]
Hi, I have my certificates accessible from the file system on the docker host so it would be great when adding a custom certificate that you could specify a local path instead of manually uploading the required files,

That way I could setup the container with e.g.:

    volumes
      - /var/ssl:/ssl

And when adding a custom certificate I could enter e.g. `/ssl/my.cert`

**Comment by vrelk** [comment]
This would be very helpful. I use a .net for my local network, so being able to use something such as acme.sh to generate a wildcard using a txt dns entry would be great and not have to copy the cert over manually.

For reference: https://github.com/Neilpang/acme.sh

**Comment by cbrherms** [comment]
I would also like this. Would be useful for wildcard certs i'm generating via a different container.

**Comment by Dragonpark** [comment]
I would also love this feature as this is the only thing preventing me from using this. I also use a different container to generate my certs (via DNS, not HTTP), which also handles automatically restarting any required containers when the certificate updates.

**Comment by ubergeek77** [comment]
Apologies for the ping, but can we get a comment on this, @jc21 ?

**Comment by osnet** [comment]
the actual compose magic is putting the certs persistent in  ./letsencrypt/ 
![image](https://user-images.githubusercontent.com/1938728/101133522-2cd9aa00-3609-11eb-9ad8-175f00128842.png)
![image](https://user-images.githubusercontent.com/1938728/101133552-382cd580-3609-11eb-9a4a-34a1fea6f773.png)
![image](https://user-images.githubusercontent.com/1938728/101133570-43800100-3609-11eb-9931-702ec3c2f0fc.png)

ok you have to know what npm-ID is your  cert  but  thats ok in smaller environments

**Comment by christian-skjetne** [comment]
There is a way to do this using the web interface. You can just turn off the SSL setting in the "proxy host" settings (set to none) and in the "Advanced"-tab you can set the ssl nginx manually using "Custom Nginx Configuration". 
Example:
```
listen 443 ssl http2;
listen [::]:443;

  # Let's Encrypt SSL
  include conf.d/include/letsencrypt-acme-challenge.conf;
  include conf.d/include/ssl-ciphers.conf;
  ssl_certificate /path/to/cert/cert.pem;
  ssl_certificate_key /path/to/cert/key.pem;

```
just copy and paste this for every "proxy host" you want to add. Not perfect or elegant, but it seems to work.

**Comment by rubendepuben** [comment]
Damn this feature would be nice

**Comment by meinradr** [comment]
This feature would be amazing. Specially for non proxy host certificates since there is no advanced tab and therefore @christian-skjetne workaround does not work.

**Comment by shalak** [comment]
Any updates on this one? I'm managing certbot separately and I'm provisioning all my services with certs via hooks. I'd like to see a "load cert from this path" option in npm...

**Comment by azoller1** [comment]
@shalak 

Does @christian-skjetne comment not help you? Should be ok for any http proxy route. Haven't tried this out yet, though. Really want to use NPM, but this is a feature I would like to have. But, I think this should work for now.

**Comment by shalak** [comment]
> Does @christian-skjetne comment not help you? 

Yes, it does solve the issue, but I do not like to have such workarounds. If I have to manage configuration manually, I prefer to keep it all manually and not use NPM at all.

**Comment by BigW72** [comment]
> There is a way to do this using the web interface. You can just turn off the SSL setting in the "proxy host" settings (set to none) and in the "Advanced"-tab you can set the ssl nginx manually using "Custom Nginx Configuration". Example:
> 
> ...
> 
Apologies if this is an idiot question, but why does it not seem to work to stop the container, replace the `privkey.pem` and `fullchain.pem` in the relevant `/data/custom_ssl/npm-x` folder with manually updated cert/key files and restarting the container?

**Comment by heiybb** [comment]
> > There is a way to do this using the web interface. You can just turn off the SSL setting in the "proxy host" settings (set to none) and in the "Advanced"-tab you can set the ssl nginx manually using "Custom Nginx Configuration". Example:
> > ...
> 
> Apologies if this is an idiot question, but why does it not seem to work to stop the container, replace the `privkey.pem` and `fullchain.pem` in the relevant `/data/custom_ssl/npm-x` folder with manually updated cert/key files and restarting the container?

Maybe cuz the certificate content is actually stored in the db.
If you open the database.sqlite and have a check with the `certificate` table you will find that they store the cert as json format in it.

**Comment by heiybb** [comment]
> > > There is a way to do this using the web interface. You can just turn off the SSL setting in the "proxy host" settings (set to none) and in the "Advanced"-tab you can set the ssl nginx manually using "Custom Nginx Configuration". Example:
> > > ...
> > 
> > 
> > Apologies if this is an idiot question, but why does it not seem to work to stop the container, replace the `privkey.pem` and `fullchain.pem` in the relevant `/data/custom_ssl/npm-x` folder with manually updated cert/key files and restarting the container?
> 
> The certificate content is actually stored in the db. If you open the `database.sqlite` and have a check with the `certificate` table you will find that they store the cert as json format in it. Maybe they read the cert from db rather than directly from the file?

Yeah ok they did read cert from db, the `proxy_host` table's cert field links to the `certificate` id.
Will try if replacing the cert content in table works or not.

**Comment by 1liminal1** [comment]
Really LOL, I mean, great project but what about those of us that want to add our own certs. I cant believe this is still a thing HAHA 

I have tried the workaround above, but get; 

Internal Error :(

**Comment by 1liminal1** [comment]
> There is a way to do this using the web interface. You can just turn off the SSL setting in the "proxy host" settings (set to none) and in the "Advanced"-tab you can set the ssl nginx manually using "Custom Nginx Configuration". Example:
> 
> ```
> listen 443 ssl http2;
> listen [::]:443;
> 
>   # Let's Encrypt SSL
>   include conf.d/include/letsencrypt-acme-challenge.conf;
>   include conf.d/include/ssl-ciphers.conf;
>   ssl_certificate /path/to/cert/cert.pem;
>   ssl_certificate_key /path/to/cert/key.pem;
> ```
> 
> just copy and paste this for every "proxy host" you want to add. Not perfect or elegant, but it seems to work.
> 
> P.S: to force ssl redirect add the line `include conf.d/include/force-ssl.conf;`

Can anyone advise what Im doing wrong here? 

I add the config as specified and it gives me an offline error? Essentially breaks it 

![image](https://user-images.githubusercontent.com/10505942/195568301-dcf268b0-f8ba-4fa9-b1fc-8d3cbdef80a1.png)

**Comment by tuxthepenguin84** [comment]
@1liminal1 
I had a similar issue until I realized I had a conflicting config in it (listen 80 was in "details" and "advanced", once I removed it from "advanced" it worked). I would double check all your paths are correct and possibly restart the container, here's my exact config:

```
listen 443 ssl http2;
include conf.d/include/ssl-ciphers.conf;
include conf.d/include/force-ssl.conf;
ssl_certificate /data/custom_ssl/fullchain.pem;
ssl_certificate_key /data/custom_ssl/privkey.pem;
```

**Comment by BigW72** [comment]
> There's a simpler way to do this if all your Proxy Host need the same SSL config.

Great, thanks! I have also tested this with a single proxy host and it works - now I can update the SSL settings for the 40+ others for the last time!

**Comment by ThaDaVos** [comment]
Is there any progress on this issue except the mentioned workaround?

**Comment by Ducksoul** [comment]
I had to declare absolute paths within the include lines. Otherwise the proxy hosts went offline:
`include /etc/nginx/conf.d/include/letsencrypt-acme-challenge.conf;`

**Comment by nsaneadmin** [comment]
> I had to declare absolute paths within the include lines. Otherwise the proxy hosts went offline: `include /etc/nginx/conf.d/include/letsencrypt-acme-challenge.conf;`

full path also fixed mine! Good find Thanks!

**Comment by shalafi99** [comment]
I too would be interested if this feature were to be included natively in NPM (thus eliminating the need to use the "Advanced" tab in each proxy host or going with the /data/nginx/custom/server_proxy.conf customization)

**Comment by cwilson1776** [comment]
Regarding [tuxthepenguin84](https://github.com/tuxthepenguin84)'s [advice](https://github.com/NginxProxyManager/nginx-proxy-manager/issues/87#issuecomment-1324162616)...

If your network and docker environment is set up for dual-stack, you need to also enable port 443 on IPv6:

```
listen 443 ssl http2;
listen [::]:443 ssl http2;
include conf.d/include/ssl-ciphers.conf;
include conf.d/include/force-ssl.conf;
ssl_certificate /data/custom_ssl/fullchain.pem;
ssl_certificate_key /data/custom_ssl/privkey.pem;
```

or you may get the dreaded SSL_ERROR_UNRECOGNIZED_NAME_ALERT error when connecting. (What's terrible about this error response is that the nginx logs will not report issues during the SSL negotiation phase, so...you've got nothing to go on when trying to debug).

**Comment by AngelGonePro** [comment]
Non of this is working for me
![image](https://github.com/user-attachments/assets/7532cbe8-6642-40d3-a0c8-78d658551a6c)

**Comment by AngelGonePro** [comment]
@tuxthepenguin84 Your method doesn't work. I just get a Unable to connect(to X.X.X.X:81) after restarting the docker. I have multiple domains so one file will not work for the multiple domains.
This feature needs to be added in the SSL section instead of having to upload them every renewal when certbot renews them.

**Comment by slovely** [comment]
> Non of this is working for me Please for the love of God add an option to add custom located certs, this is annoying and aggravating. Your own cert system gets errors, so having to go this route is my only option. ![image](https://github.com/user-attachments/assets/7532cbe8-6642-40d3-a0c8-78d658551a6c)

I suspect you've just hit the same issue I have.  If you look in `/data/logs/fallback_error.log` my guess is you'll see something like:
```
the "listen ... http2" directive is deprecated, use the "http2" directive instead in /data/nginx/proxy_host/5.conf:31
```
The later version of nginx has moved how http2 is specified (see https://forum.hestiacp.com/t/nginx-1-25-1-listen-http2-directive-is-deprecated/9816).  Now http2 should be removed from the `listen` line and moved into it's own line below as `http2 on;`.

HOWEVER - because on the SSL tab you are not selecting a cert, then you cannot enable http2 so NPM is already adding a conflicting `http2 off;` line...

Can we get the option to enable HTTP/2 support _without_ having to choose a certificate please?

**Comment by tuxthepenguin84** [comment]
For everyone having issues it might be better to move to pure nginx. I did that a while back and haven't had any further problems.

**Comment by AngelGonePro** [comment]
> > Non of this is working for me Please for the love of God add an option to add custom located certs, this is annoying and aggravating. Your own cert system gets errors, so having to go this route is my only option. ![image](https://github.com/user-attachments/assets/7532cbe8-6642-40d3-a0c8-78d658551a6c)
> 
> I suspect you've just hit the same issue I have. If you look in `/data/logs/fallback_error.log` my guess is you'll see something like:
> 
> ```
> the "listen ... http2" directive is deprecated, use the "http2" directive instead in /data/nginx/proxy_host/5.conf:31
> ```
> 
> The later version of nginx has moved how http2 is specified (see https://forum.hestiacp.com/t/nginx-1-25-1-listen-http2-directive-is-deprecated/9816). Now http2 should be removed from the `listen` line and moved into it's own line below as `http2 on;`.
> 
> HOWEVER - because on the SSL tab you are not selecting a cert, then you cannot enable http2 so NPM is already adding a conflicting `http2 off;` line...
> 
> Can we get the option to enable HTTP/2 support _without_ having to choose a certificate please?

No, my log was that NPM couldn't somehow read the file, said `was expecting  --- PUBLIC KEY---` or something on the lies of that. Even though the fike was generated using certbot and was generated correctly.

**Comment by bartgrefte** [comment]
> Issue is now considered stale. If you want to keep it open, please comment 👍

Well, then here's a comment to keep it open ;)

**Comment by BigW72** [comment]
> Issue is now considered stale. If you want to keep it open, please comment :+1:

I still think it would be great if one could easily use externally managed certificates.

If you can't use the GUI for everything and have to manage text files without certainty that the workaround won't stop working then using something designed to be managed via config files works better for me.

I have recently switched to SWAG and miss the NPM GUI, but overall it is easier to add new hosts.

PS: The reason I had to manage certs externally is because my DNS doesn't support automated renewal.

**Comment by BeanBagBob** [comment]
I would love for this to be added to NPM natively.

**Comment by BeanBagBob** [comment]
Question. Will NPM notice when the certificate on the host has been renewed?

**Comment by smolurl** [comment]
Re-up please

**Comment by MK796** [comment]
Please keep that open 🙂

**Comment by bkrajendra** [comment]
Can we have a custom ssl auto renew automation to be integrated in the nginxproxymanager.
Im now sure how this can be done, but i heavyly use the this for frontending my local lan application with mydomain.local.
And i have my own custom root CA cert and I had to generate new SSL cert every year, keep tracking of its expiration manualy.
I do have scripts,. but it would be greate to have something like letsEncrypt for local custom certs integrated into nginxproxymanager.

Thanks in advance.
